### PR TITLE
fix: HEAD route reseting

### LINKED ISSUE #5089: Method 'HEAD' already declared for route - regression in 4.24.0
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.24.0

### Plugin version

_No response_

### Node.js version

16.20

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

latest

### Description

With the 4.24.0 release, our application no longer starts due to an error:

```
Failed to start: Error: Method 'HEAD' already declared for route '/api/v1/devices/:/editor/proxy/*' with constraints '{}'
```

Our application makes two calls to `app.route` - the first handles GET requests that *could* be a WebSocket connection, the second handles all the other methods we need to cope with.

```
    app.route({
        method: 'GET',
        url: '/proxy/*',
        ...
    })

    app.route({
        method: ['POST', 'DELETE', 'PUT', 'HEAD', 'OPTIONS'],
        url: '/proxy/*',
        ...
    })
```

With 4.24.0, this second call is triggering the error. Previous versions worked fine.

### Steps to Reproduce

I will update with a link to a repo containing as minimal reproduction as I can create tomorrow.

I wanted to get the issue recorded quickly as it appears to be a regression in 4.24.0. Looking at PRs in the release, https://github.com/fastify/fastify/pull/5078 appears very relevant.

### Expected Behavior

_No response_

### PR DESCRIPTION
**Pull request by ivan-tymoshenko**
Fix #5089

### CODE REVIEW

File: lib/route.js [hunk 1]
```diff
@@ -327,7 +327,7 @@ function buildRouting (options) {
 
       // remove the head route created by fastify
       if (hasHEADHandler && !context[kRouteByFastify] && headHandler.store[kRouteByFastify]) {
-        router.off('HEAD', opts.url, { constraints })
+        router.off('HEAD', opts.url, constraints)
```
Comment (climba03003) [inline lib/route.js:330]: Why `.on` and `.off` is not using the same syntax?

**Review by climba03003 — CHANGES_REQUESTED** [review]

**Comment by climba03003** [comment]
Some test need to be adjusted.

**Comment by ivan-tymoshenko** [comment]
> Some test need to be adjusted.

What tests?

**Comment by climba03003** [comment]
```
    # Subtest: pretty print - commonPrefix
        1..4
        ok 1 - should be equal
        ok 2 - should be equal
        not ok 3 - should be equal
          ---
          compare: ===
          at:
            line: 268
            column: 7
            file: test/pretty-print.test.js
          stack: |
            test/pretty-print.test.js:268:7
            manageErr (fastify.js:602:11)
            exit (lib/hooks.js:113:5)
            next (lib/hooks.js:124:9)
            hookRunnerApplication (lib/hooks.js:98:3)
            fastify.js:584:11
            Object._encapsulateThreeParam (node_modules/avvio/boot.js:562:7)
            Boot.timeoutCall (node_modules/avvio/boot.js:458:5)
            Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:440:19)
            release (node_modules/fastq/queue.js:149:16)
            Object.resume (node_modules/fastq/queue.js:82:7)
            node_modules/avvio/boot.js:174:18
            node_modules/avvio/plugin.js:275:7
            done (node_modules/avvio/plugin.js:200:5)
            check (node_modules/avvio/plugin.js:224:9)
          source: |2
                t.equal(typeof flatTree, 'string')
                t.equal(radixTree, radixExpected)
            ------^
                t.equal(flatTree, flatExpected)
              })
          diff: |+
            --- expected
            +++ actual
            @@ -1,5 +1,5 @@
             └── /
                 └── hel
            -        ├── lo (GET, HEAD, PUT)
            +        ├── lo (GET, PUT)
                     └── icopter (GET, HEAD)
          
          ...
        
        not ok 4 - should be equal
          ---
          compare: ===
          at:
            line: 269
            column: 7
            file: test/pretty-print.test.js
          stack: |
            test/pretty-print.test.js:269:7
            manageErr (fastify.js:602:11)
            exit (lib/hooks.js:113:5)
            next (lib/hooks.js:124:9)
            hookRunnerApplication (lib/hooks.js:98:3)
            fastify.js:584:11
            Object._encapsulateThreeParam (node_modules/avvio/boot.js:562:7)
            Boot.timeoutCall (node_modules/avvio/boot.js:458:5)
            Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:440:19)
            release (node_modules/fastq/queue.js:149:16)
            Object.resume (node_modules/fastq/queue.js:82:7)
            node_modules/avvio/boot.js:174:18
            node_modules/avvio/plugin.js:275:7
            done (node_modules/avvio/plugin.js:200:5)
            check (node_modules/avvio/plugin.js:224:9)
          source: |2
                t.equal(radixTree, radixExpected)
                t.equal(flatTree, flatExpected)
            ------^
              })
            })
          diff: |+
            --- expected
            +++ actual
            @@ -1,3 +1,3 @@
            -├── /hello (GET, HEAD, PUT)
            +├── /hello (GET, PUT)
             └── /helicopter (GET, HEAD)
          
          ...
        
        # failed 2 of 4 tests
    not ok 10 - pretty print - commonPrefix # time=77.061ms
    
    # Subtest: pretty print - includeMeta, includeHooks
        1..6
        ok 1 - should be equal
        ok 2 - should be equal
        ok 3 - should be equal
        not ok 4 - should be equal
          ---
          compare: ===
          at:
            line: 352
            column: 7
            file: test/pretty-print.test.js
          stack: |
            test/pretty-print.test.js:352:7
            manageErr (fastify.js:602:11)
            exit (lib/hooks.js:113:5)
            next (lib/hooks.js:124:9)
            hookRunnerApplication (lib/hooks.js:98:3)
            fastify.js:584:11
            Object._encapsulateThreeParam (node_modules/avvio/boot.js:562:7)
            Boot.timeoutCall (node_modules/avvio/boot.js:458:5)
            Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:440:19)
            release (node_modules/fastq/queue.js:149:16)
            Object.resume (node_modules/fastq/queue.js:82:7)
            node_modules/avvio/boot.js:174:18
            node_modules/avvio/plugin.js:275:7
            done (node_modules/avvio/plugin.js:200:5)
            check (node_modules/avvio/plugin.js:224:9)
          source: |2
                t.equal(typeof hooksOnlyExpected, 'string')
                t.equal(radixTree, radixExpected)
            ------^
                t.equal(flatTree, flatExpected)
                t.equal(hooksOnly, hooksOnlyExpected)
          diff: |
            --- expected
            +++ actual
            @@ -3,13 +3,8 @@
                     ├── lo (GET, PUT)
                     │   • (onTimeout) ["onTimeout()"]
                     │   • (onRequest) ["anonymous()"]
                     │   • (errorHandler) "defaultErrorHandler()"
            -        │   lo (HEAD)
            -        │   • (onTimeout) ["onTimeout()"]
            -        │   • (onRequest) ["anonymous()"]
            -        │   • (onSend) ["headRouteOnSendHandler()"]
            -        │   • (errorHandler) "defaultErrorHandler()"
                     └── icopter (GET)
                         • (onTimeout) ["onTimeout()"]
                         • (onRequest) ["anonymous()"]
                         • (errorHandler) "defaultErrorHandler()"
          ...
        
        not ok 5 - should be equal
          ---
          compare: ===
          at:
            line: 353
            column: 7
            file: test/pretty-print.test.js
          stack: |
            test/pretty-print.test.js:353:7
            manageErr (fastify.js:602:11)
            exit (lib/hooks.js:113:5)
            next (lib/hooks.js:124:9)
            hookRunnerApplication (lib/hooks.js:98:3)
            fastify.js:584:11
            Object._encapsulateThreeParam (node_modules/avvio/boot.js:562:7)
            Boot.timeoutCall (node_modules/avvio/boot.js:458:5)
            Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:440:19)
            release (node_modules/fastq/queue.js:149:16)
            Object.resume (node_modules/fastq/queue.js:82:7)
            node_modules/avvio/boot.js:174:18
            node_modules/avvio/plugin.js:275:7
            done (node_modules/avvio/plugin.js:200:5)
            check (node_modules/avvio/plugin.js:224:9)
          source: |2
                t.equal(radixTree, radixExpected)
                t.equal(flatTree, flatExpected)
            ------^
                t.equal(hooksOnly, hooksOnlyExpected)
              })
          diff: |
            --- expected
            +++ actual
            @@ -1,13 +1,8 @@
             ├── /hello (GET, PUT)
             │   • (onTimeout) ["onTimeout()"]
             │   • (onRequest) ["anonymous()"]
             │   • (errorHandler) "defaultErrorHandler()"
            -│   /hello (HEAD)
            -│   • (onTimeout) ["onTimeout()"]
            -│   • (onRequest) ["anonymous()"]
            -│   • (onSend) ["headRouteOnSendHandler()"]
            -│   • (errorHandler) "defaultErrorHandler()"
             └── /helicopter (GET)
                 • (onTimeout) ["onTimeout()"]
                 • (onRequest) ["anonymous()"]
                 • (errorHandler) "defaultErrorHandler()"
          ...
        
        not ok 6 - should be equal
          ---
          compare: ===
          at:
            line: 354
            column: 7
            file: test/pretty-print.test.js
          stack: |
            test/pretty-print.test.js:354:7
            manageErr (fastify.js:602:11)
            exit (lib/hooks.js:113:5)
            next (lib/hooks.js:124:9)
            hookRunnerApplication (lib/hooks.js:98:3)
            fastify.js:584:11
            Object._encapsulateThreeParam (node_modules/avvio/boot.js:562:7)
            Boot.timeoutCall (node_modules/avvio/boot.js:458:5)
            Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:440:19)
            release (node_modules/fastq/queue.js:149:16)
            Object.resume (node_modules/fastq/queue.js:82:7)
            node_modules/avvio/boot.js:174:18
            node_modules/avvio/plugin.js:275:7
            done (node_modules/avvio/plugin.js:200:5)
            check (node_modules/avvio/plugin.js:224:9)
          source: |2
                t.equal(flatTree, flatExpected)
                t.equal(hooksOnly, hooksOnlyExpected)
            ------^
              })
            })
          diff: |
            --- expected
            +++ actual
            @@ -1,11 +1,7 @@
             ├── /hello (GET, PUT)
             │   • (onTimeout) ["onTimeout()"]
             │   • (onRequest) ["anonymous()"]
            -│   /hello (HEAD)
            -│   • (onTimeout) ["onTimeout()"]
            -│   • (onRequest) ["anonymous()"]
            -│   • (onSend) ["headRouteOnSendHandler()"]
             └── /helicopter (GET)
                 • (onTimeout) ["onTimeout()"]
                 • (onRequest) ["anonymous()"]
                 /helicopter (HEAD)
          ...
        
        # failed 3 of 6 tests
    not ok 11 - pretty print - includeMeta, includeHooks # time=24.811ms
    
    1..11
    # failed 2 of 11 tests
    # time=210.844ms
not ok 100 - test/pretty-print.test.js # time=210.844ms
```

**Comment by ivan-tymoshenko** [comment]
Can't tell know, but having that this tests shouldn't fail, it looks like another bug in fastify.

**Comment by ivan-tymoshenko** [comment]
I didn't expect it to be broken on so many levels.

**Review by mcollina — APPROVED** [review]
lgtm

**Review by climba03003 — APPROVED** [review]
The logic looks fine.

**Pull request merged**